privacyidea/lib/tokens/vascotoken.py

```diff
@@ -0,0 +1,162 @@
+# License as published by the Free Software Foundation; either
+# version 3 of the License, or any later version.
+#
+# This code is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU AFFERO GENERAL PUBLIC LICENSE for more details.
+#
+# You should have received a copy of the GNU Affero General Public
+# License along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+#
+import binascii
+
+__doc__ = """This is the implementation of the VASCO token"""
+
+import logging
+from privacyidea.api.lib.utils import getParam
+from privacyidea.lib.utils import is_true
+from privacyidea.lib.decorators import check_token_locked
+from privacyidea.lib.error import ParameterError
+from privacyidea.lib.log import log_with
+from privacyidea.lib.tokenclass import TokenClass
+from privacyidea.lib.tokens.vasco import vasco_otp_check
+from privacyidea.lib import _
+
+optional = True
+required = False
+
+log = logging.getLogger(__name__)
+
+
+class VascoTokenClass(TokenClass):
+    """
+    Token class for VASCO Digipass tokens. Relies on vendor-specific
+    shared library, whose location needs to be set in the VASCO_LIBRARY
+    config option.
+
+    VASCO Tokens can be read from a CSV file which is structured as follows::
+
+        <serial1>,<hexlify(blob1)>,vasco
+        <serial2>,<hexlify(blob2)>,vasco
+        ...
+
+    whereas blobX is the 248-byte blob holding the token information.
+    Consequently, hexlify(blobX) is a 496-character hex string.
+
+    The CSV file can be imported by using the "Import Tokens" feature of the Web UI,
+    where "OATH CSV" needs to be chosen as the file type.
+    """
+
+    def __init__(self, db_token):
+        """
+        constructor - create a token class object with its db token binding
+
+        :param aToken: the db bound token
+        """
+        TokenClass.__init__(self, db_token)
+        self.set_type(u"vasco")
+        self.hKeyRequired = True
+
+    @staticmethod
+    def get_class_type():
+        """
+        return the class type identifier
+        """
+        return "vasco"
+
+    @staticmethod
+    def get_class_prefix():
+        """
+        return the token type prefix
+        """
+        return "VASC"
+
+    @staticmethod
+    @log_with(log)
+    def get_class_info(key=None, ret='all'):
+        """
+        :param key: subsection identifier
+        :type key: string
+        :param ret: default return value, if nothing is found
+        :type ret: user defined
+        :return: subsection if key exists or user defined
+        :rtype: dict or string
+        """
+        res = {'type': 'vasco',
+               'title': 'VASCO Token',
+               'description': _('VASCO Token: Authentication using VASCO tokens'),
+               'user': ["enroll"],
+               'ui_enroll': ["admin", "user"],
```

Comment:
Do we really want allow a ui_enrollment? I.e. the admin or the user would need to paste the hexlified blob into the webui?
I think this probably never will happen. But anyway, this looks cool in the UI, to have an additional token type;-)

Reply:
I noticed that because we defined no template, VASCO tokens cannot be enrolled at all (because we cannot paste a blob as the OTP key). So in 53b6bf8, I removed the ``ui_enroll`` line -- what do you think?